pkg/analyze/file.go

```diff
@@ -3,13 +3,59 @@ package analyze
 import (
 	"io"
 	"os"
+	"os/exec"
+	"path/filepath"
 
 	"github.com/nxadm/tail"
 	"github.com/taoky/ayano/pkg/fileiter"
 )
 
 const oneMiB = 1024 * 1024
 
+func filterByCommand(r io.Reader, args []string) (io.ReadCloser, error) {
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdin = r
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	go func() {
+		cmd.Wait()
+		if closer, ok := r.(io.Closer); ok {
+			closer.Close()
+		}
+	}()
+	return stdout, err
+}
+
+type filterFunc func(r io.Reader) (io.Reader, error)
+
+var fileTypes = map[string]filterFunc{
+	".gz": func(r io.Reader) (io.Reader, error) {
+		return filterByCommand(r, []string{"gzip", "-cd"})
+	},
+	".xz": func(r io.Reader) (io.Reader, error) {
+		return filterByCommand(r, []string{"xz", "-cd", "-T", "0"})
```

Comment:
Do you think that we shall let users choose how many threads that ayano shall use?

Reply:
man zstd(1):

```
       •   -T#, --threads=#: Compress using # working threads (default: 1). If # is 0, attempt to detect and use the number of  physi‐
           cal CPU cores. In all cases, the nb of threads is capped to ZSTDMT_NBWORKERS_MAX, which is either 64 in 32-bit mode, or 256
           for 64-bit environments. This modifier does nothing if zstd is compiled without multithread support.
```

man xz(1):

```
       -T threads, --threads=threads
              Specify the number of worker threads to use.  Setting threads to a special value 0 makes xz use up to as many threads as
              the processor(s) on the system support.  The actual number of threads can be fewer than threads if the input file is not
              big enough for threading with the given settings or if using more threads would exceed the memory usage limit.

...

              The default value for threads is 0.  In xz 5.4.x and older the default is 1.
```

I guess it's fine not to implement it ourselves now.